stl/inc/xutility

```diff
@@ -4313,72 +4313,116 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+    is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source&>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+template <class _Ty>
+_INLINE_VAR constexpr bool _Can_memop_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memop_cat::_Move, false> =
+    is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<move_iterator<_Source*>, _Dest*> : _Ptr_copy_cat<_Source*, _Dest*> {};
+// Types are eligible for memcpy / memmove into uninitialized memory when they are trivial.
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memop_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memop_cat::_Copy_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_move_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for move
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memop_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memop_cat::_Move_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
+// Different types are eligible for memcpy when they are trivially copy assignable.
+template <class _Source, class _Dest, _Memop_cat>
+_INLINE_VAR constexpr bool _Can_assign_elements = is_trivially_assignable_v<_Dest&, _Source&>;
+
+// Different types are eligible for memmove when they are trivially move assignable.
 template <class _Source, class _Dest>
-struct _Ptr_move_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+_INLINE_VAR constexpr bool _Can_assign_elements<_Source, _Dest, _Memop_cat::_Move> =
+    is_trivially_assignable_v<_Dest&, _Source&&>;
 
 template <class _Source, class _Dest>
-struct _Ptr_move_cat<move_iterator<_Source*>, _Dest*> : _Ptr_move_cat<_Source*, _Dest*> {};
+_INLINE_VAR constexpr bool _Can_assign_elements<_Source, _Dest, _Memop_cat::_Move_uninitialized> =
+    is_trivially_assignable_v<_Dest&, _Source&&>;
+
+// _Can_memop<_IterSource, _IterDest> reports whether we can activate the memory operation for
+// arbitrary iterators. It ignores top-level constness on the source iterators and on the source elements.
+template <class _IterSource, class _IterDest, _Memop_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memop = false;
+
+template <class _Source, class _Dest, _Memop_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memop<_Source*, _Dest*, _Cat> =
+    conjunction_v<bool_constant<_Can_assign_elements<_Source, _Dest, _Cat>>,
+        bool_constant<_Can_memop_elements<remove_cv_t<_Source>, remove_volatile_t<_Dest>, _Cat>>>;
+
+template <class _IterSource, class _IterDest, _Memop_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memop<move_iterator<_IterSource>, _IterDest, _Cat> =

```

Comment:
Shouldn't `_Copy` and `_Copy_uninitialized` become `_Move` and `_Move_uninitialized` with `move_iterator`? (I know it's pre-existing but it's definitely wrong)

Reply:
I think that you are right that for the `_Copy` case it seems wrong. For `_Copy_uninitialized` we require `trivial` types anyway where it does not make a difference.

Thatsaid I am absolutely in favor of changing it accordingly

Reply:
Since we fixed it in #1772 it's no longer pre-existing so we should change this accordingly.